### PR TITLE
PARQUET-353: Recycle compressors in parquet write path

### PR DESCRIPTION
**Pull request by nitin2goyal**

### CODE REVIEW

**Comment by liancheng** [comment]
How can we retest this PR? The last failure should have been fixed by #269.

**Comment by nitin2goyal** [comment]
Commited a dummy checkin such that tests re-run automatically

**Comment by liancheng** [comment]
+1

@nitin2goyal Thanks for the fix :)

**Comment by julienledem** [comment]
This code has changed. you may want to rebase.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileWriter.java [hunk 1]
```diff
@@ -171,6 +171,8 @@ private final STATE error() throws IOException {
 
   private STATE state = STATE.NOT_STARTED;
 
+  private final CodecFactory codecFactory;
```
Comment (julienledem) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileWriter.java]: It seems that this is not the right place to put this as this class does not use the codecFactory.
it just creates it and makes it available. I'd rather have private members not accessible to the outside to respect encapsulation and layering of the code.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetOutputFormat.java [hunk 2]
```diff
@@ -341,7 +341,6 @@ private static int getMaxPaddingSize(Configuration conf) {
         throws IOException, InterruptedException {
     final WriteSupport<T> writeSupport = getWriteSupport(conf);
 
-    CodecFactory codecFactory = new CodecFactory(conf);
```
Comment (julienledem) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetOutputFormat.java]: I don't get this change. the creation has been moved to ParquetFileWriter w = new ParquetFileWriter(... just a few lines bellow. Could you explain why we do this?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetWriter.java [hunk 3]
```diff
@@ -267,8 +267,7 @@ public ParquetWriter(Path file, Configuration conf, WriteSupport<T> writeSupport
         conf, schema, file, mode, blockSize, maxPaddingSize);
     fileWriter.start();
 
-    CodecFactory codecFactory = new CodecFactory(conf);
-    CodecFactory.BytesCompressor compressor =	codecFactory.getCompressor(compressionCodecName, 0);
+    CodecFactory.BytesCompressor compressor =	fileWriter.getCodecFactory().getCompressor(compressionCodecName, 0);
```
Comment (julienledem) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetWriter.java]: possibly what you really want is pass the codec factory as a parameter to this constructor.

**Comment by julienledem** [comment]
Thanks for looking into this. A lot of the codec factory code has changed recently.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileWriter.java (hunk 1 above)
Comment (rdblue) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileWriter.java]: +1

**Comment by rdblue** [comment]
It looks like the purpose of this is to tie the lifecycle of the CodecFactory to ParquetFileWriter. That seems like a good idea (resources are freed when the file is closed), but ends up exposing internals (as Julien noted) because we don't delegate compression to ParquetFileWriter.

I think it would be cleaner to attach the lifecycle of the codec factory to the record writer for MR, which has a close method for clean up. The CodecFactory would be instantiated in the writer's constructor instead of passing in a compressor. This should also call `release` in the `ParquetWriter#close` method to avoid the same leak when not writing via MR.

**Comment by rdblue** [comment]
To get this into the pending release, I've implemented the changes I recommended in #295.

**Pull request closed without merging**